packages/opentelemetry-api/src/diag/logger.ts

```diff
@@ -0,0 +1,158 @@
+ * that are currently passing Logger references during migration to minimize
+ * breaks that will occur with the removal of the Logger interface.
+ */
+export type OptionalDiagLogger = Logger | DiagLogger | null | undefined;
+
+export type DiagLogFunction = (message: string, ...args: unknown[]) => void;
+
+/** Defines an internal diagnostic logger interface which is used to log internal diagnostic
+ * messages, you can set the default diagnostic logger via the {@link DiagAPI} setLogger function.
+ * API provided implementations include :-
+ * - a No-Op {@link noopDiagLogger}
+ * - a {@link DiagLogLevel} filtering wrapper {@link diagLogLevelFilter}
+ * - a general Console {@link DiagConsoleLogger} version.
+ */
+export interface DiagLogger {
+  /** Log a terminal situation that would cause the API to completely fail to initialize,
+   * if this type of message is logged functionality of the API is not expected to be functional.
+   */
+  terminal: DiagLogFunction;
+
+  /** Log a critical error that NEEDS to be addressed, functionality of the component that emits
+   * this log detail may non-functional. While the overall API may be.
+   */
+  critical: DiagLogFunction;
+
+  /** Log an error scenario that was not expected and caused the requested operation to fail. */
+  error: DiagLogFunction;
+
+  /** Log a warning scenario to inform the developer of an issues that should be investigated.
+   * The requested operation may or may not have succeeded or completed.
+   */
+  warn: DiagLogFunction;
+
+  /** Log a general informational message, this should not affect functionality.
+   * This is also the default logging level so this should NOT be used for logging
+   * debugging level information.
+   */
+  info: DiagLogFunction;
+
+  /** Log a general debug message that can be useful for identifying a failure.
+   * Information logged at this level may include diagnostic details that would
+   * help identify a failure scenario. Useful scenarios would be to log the execution
+   * order of async operations
+   */
+  debug: DiagLogFunction;
+
+  /** Log a detailed (verbose) trace level logging that can be used to identify failures
+   * where debug level logging would be insufficient, this level of tracing can include
+   * input and output parameters and as such may include PII information passing through
+   * the API. As such it is recommended that this level of tracing should not be enabled
+   * in a production environment.
+   */
+  trace: DiagLogFunction;
+
+  /** Log a general informational message that should always be logged regardless of the
+   * current {@Link DiagLogLevel) and configured filtering level. This type of logging is
+   * useful for logging component startup and version information without causing additional
+   * general informational messages when the logging level is set to DiagLogLevel.WARN or lower.
+   */
+  forcedInfo: DiagLogFunction;
+}
+
+const loggerMapFuncs: { d: keyof DiagLogger; l: keyof Logger }[] = [
```

Comment:
This will be removed as part of the Logger removal, as this is a mapping from the DiagLogger functions to the Logger versions.